dali/operators/reader/numpy_reader_gpu_op.cc

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "dali/pipeline/data/views.h"
+#include "dali/kernels/transpose/transpose_gpu.h"
+#include "dali/core/static_switch.h"
+#include "dali/operators/reader/numpy_reader_gpu_op.h"
+
+namespace dali {
+
+void NumpyReaderGPU::Prefetch() {
+  // We actually prepare the next batch
+  DomainTimeRange tr("[DALI][NumpyReaderGPU] Prefetch #" + to_string(curr_batch_producer_),
+                      DomainTimeRange::kRed);
+  DataReader<GPUBackend, ImageFileWrapperGPU>::Prefetch();
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+  auto &curr_tensor_list = prefetched_batch_tensors_[curr_batch_producer_];
+
+  // get shapes
+  for (size_t data_idx = 0; data_idx < curr_batch.size(); ++data_idx) {
+    thread_pool_.AddWork([&curr_batch, data_idx](int tid) {
+        curr_batch[data_idx]->read_meta_f();
+      });
+  }
+  thread_pool_.RunAll();
```

Comment:
I wonder how it would affect us if the CPU work was already done by the loader and stored in the prefetched_batch_queue_, so we can proceed with reading via GDS.

Reply:
I don't understand.